### PR TITLE
Improved unicorn engine

### PR DESCRIPTION
**Pull request by dnivra**
Rewrote the unicorn engine to speed up emulation. Most of the changes(https://github.com/angr/angr/compare/master...393f059) have been reviewed before. The following are the commits not yet reviewed:

https://github.com/angr/angr/compare/393f059...8f4231f: a bug fix and changes suggested during the last review.
https://github.com/angr/angr/commit/3665043: Native interface code refactor.
https://github.com/angr/angr/compare/3665043...feat/improved-unicorn-engine: Changes after native interface refactor. Mostly bug fixes.

### CODE REVIEW

File: angr/exploration_techniques/tracer.py [hunk 1]
```diff
@@ -495,6 +495,94 @@ def _compare_addr(self, trace_addr, state_addr):
...
             raise AngrTracerError("Trace desynced on jumping into %#x, where no library is mapped!" % state_addr)
 
+    def _check_qemu_unicorn_desync(self, state: 'angr.SimState', trace_curr_idx, state_desync_block_addr):
+        """
+        Check if desync occurred because large blocks are split up at different instructions by qemu and unicorn. This
+        is done by reconstructing part of block executed so far from the trace and state history and checking if they
+        the same
+        """
+
+        from capstone import CS_GRP_CALL, CS_GRP_IRET, CS_GRP_JUMP, CS_GRP_RET # pylint:disable=import-outside-toplevel
```
Comment (ltfish) [inline angr/exploration_techniques/tracer.py:505]: Can we move this line of import to the top of this file?

File: angr/engines/unicorn.py [hunk 2]
```diff
@@ -155,6 +238,10 @@ def process_successors(self, successors, **kwargs):
                         state._inspect('irsb', BP_AFTER, address=bbl_addr)
                     break
 
+        if state.unicorn.stop_reason in (STOP.symbolic_stop_reasons + STOP.unsupported_reasons) or \
+          state.unicorn.stop_reason in (STOP.STOP_UNKNOWN_MEMORY_WRITE, STOP.STOP_VEX_LIFT_FAILED):
```
Comment (ltfish) [inline angr/engines/unicorn.py:242]: The second line looks ugly right now. I would suggest following one of the styles in [PEP8](https://www.python.org/dev/peps/pep-0008/#multiline-if-statements).

File: angr/exploration_techniques/tracer.py [hunk 3]
```diff
@@ -495,6 +495,94 @@ def _compare_addr(self, trace_addr, state_addr):
...
+            trace_block_last_insn = trace_block.capstone.insns[-1]
+            for insn_type in control_flow_insn_types:
+                if trace_block_last_insn.group(insn_type):
+                    big_block_start_addr = self._trace[trace_block_idx + 1]
+                    break
+
+            if big_block_start_addr is not None:
+                break
+
+        if big_block_start_addr is None:
```
Comment (ltfish) [inline angr/exploration_techniques/tracer.py:528]: I would rewrite it to:

```
for ... :
    ...
else:
    # Failed to find end of the big basic block in trace. Treat as trace desync.
    return False
```

File: angr/state_plugins/unicorn_engine.py [hunk 4]
```diff
@@ -809,6 +926,52 @@ def _map_one_page(self, _uc, addr):
             _UC_NATIVE.activate(self._uc_state, start, length, taint[0] if taint else None)
             return True
 
+    def _get_details_of_blocks_with_symbolic_instrs(self):
+        def _get_register_values(register_values):
+            for register_value in register_values:
+                # Convert the register value in bytes to number of appropriate size and endianness
+                reg_name, reg_size = self.state.arch.vex_reg_offset_to_name[register_value.offset]
+                if self.state.arch.register_endness == 'Iend_LE':
```
Comment (ltfish) [inline angr/state_plugins/unicorn_engine.py:934]: I would use `archinfo.Endness.LE`, `archinfo.Endness.BE`, and `archinfo.Endness.ME`. Same for line 948.

**Review by ltfish — CHANGES_REQUESTED** [review]

File: angr/exploration_techniques/tracer.py (hunk 1 above)
Comment (dnivra) [reply angr/exploration_techniques/tracer.py]: Done.

File: angr/engines/unicorn.py (hunk 2 above)
Comment (dnivra) [reply angr/engines/unicorn.py]: Fixed.

File: angr/exploration_techniques/tracer.py (hunk 3 above)
Comment (dnivra) [reply angr/exploration_techniques/tracer.py]: Fixed.

File: angr/state_plugins/unicorn_engine.py (hunk 4 above)
Comment (dnivra) [reply angr/state_plugins/unicorn_engine.py]: Done.

**Review by ltfish — APPROVED** [review]

**Pull request merged**